refactor lockstep programming

psychocoderHPC opened the pull request:
Entirely rewrite the lockstep programming of PMacc.
The initial implementation requires that the user starts the kernel with the correct number of threads and that the kernel uses the valid number of workers (compile time) and passes the thread index correctly to the lockstep classes.
This was hard to understand and the possibility of mistakes was high.

The new implementation simplifies the usage and avoids typical errors by passing wrong indices to the lockstep classes.

# Review

I split the PR into a few commits to simplifying the review. The most critical part is PMacc cuSTL and kernel/plugins using cuSTL. The reason is that cuSTL is not maintained since years and mostly not able to handle heterogeneous kernel very well.
Many cuSTL plugins can only run on GPU :-( anyway.
NOTE: we are working on removing cuSTL but this will take some time.

- [x] contains #4337 to fix compile issues
- [x] contains #4338 to fix broken game of life

The rewrite is not increasing the register memory footprint!

- [x] update documentation
- [x] perform more runtime tests/validation
- [x] rebase against https://github.com/ComputationalRadiationPhysics/picongpu/pull/4338

psychocoderHPC commented:
Here is a simple overview of the most used transition changes in this PR.

All PMacc kernel now uses the context element instance `worker` instead of `acc` to encapsulate the information of a thread/worker and provide access to the alpaka accelerator via `worker.getAcc()`
An alpaka accelerator is only required to interact with cupla, alpaka, or mallocMC interfaces.

# A typical kernel call 

- before looked like
```C++
constexpr uint32_t numWorkers
                        = pmacc::traits::GetNumWorkers<pmacc::math::CT::volume<SuperCellSize>::type::value>::value;
// If `numWorkers` in the kernel signature and the start parameter for the number of threads differ the kernel will end in undefined behavior.
PMACC_KERNEL(KernelName<numWorkers>{})
                    (numBlocks, numWorkers)(/*kernel argumenets*/);
```

- is transformed after into
```C++
auto workerCfg = lockstep::makeWorkerCfg(SuperCellSize{});
// The kernel is not required to hand over low level information about the number of threads used in the kernel.
PMACC_LOCKSTEP_KERNEL(KernelName{}, workerCfg)(numBlocks)( /*kernel argumenets*/);
```

# A typical kernel 

- before
```C++
template<uint32_t T_numWorkers>		
struct KernelName
{
  template<typename T_Acc>
  DINLINE void operator()(T_Acc const& acc /* more arguments*/)
  {
    constexpr uint32_t numWorker = T_numWorkers;
    uint32_t const workerIdx = cupla::threadIdx(acc).x;
    // indices and number of workers must always pass correctly to the lockstep functions else the behaviour is undefined.
    auto forEachParticle = lockstep::makeForEach<frameSize, numWorker>(workerIdx);
    // ...
  }
};
```

- after
```C++	
struct KernelName
{
  template<typename T_Worker>
  DINLINE void operator()(T_Worker const& worker /* more arguments*/)
  {
    // no index mismatch possible anymore
    auto forEachParticle = lockstep::makeForEach<frameSize>(worker);
    // ...
  }
};
```

jkelling commented on include/pmacc/lockstep/WorkerCfg.hpp:
```diff
@@ -0,0 +1,117 @@
+        friend struct detail::WorkerCfgAssume1DThreads;
+
+        //! number of workers
+        static constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<T_numWorkers>::value;
```
I think the definition of `GetNumWorkers` is not (yet) changed by this PR, hence this should always result in `numWorkers = T_numWorkers`, [see here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/59e9b53605f9a5c1bf271eeb055bc74370a99052/include/pmacc/traits/GetNumWorkers.hpp#L41-L44)

psychocoderHPC replied:
No, depending on the used cupla/accelerator the trait is giving other values back. For OMP2block the trait will return`1`.
The reason is that some alpaka accelerator has a restriction for the block size.
Before this PR the user always called this trait. Now the call is moved into `WorkerCfg` because calling this trait is an essential part of lockstep programming.

jkelling replied:
Are you sure about this? The trait I linked to has a default value for the template parameter `T_ACC` and no second template parameter is passed here. To get another specialization, you would need to pass an accelerator type, e.g. `GetNumWorkers<T_numWorkers, AccOmp2Blocks<Idx, Dim>>`.

psychocoderHPC replied:
Yes I am sure.

https://github.com/ComputationalRadiationPhysics/picongpu/blob/59e9b53605f9a5c1bf271eeb055bc74370a99052/include/pmacc/traits/GetNumWorkers.hpp#L41-L52

The default will be used and then the compiler selects the specialized implementation. If this did not work you would not be able to run `serial` or `omp2b` as PIConGPU backend because both backends have the limitation that only one thread per block is allowed.

jkelling replied:
Ahh, so this is because `cupla::AccThreadSeq` is an alias for the default accelerator configured via cmake. I am not yet familiar enough with cupla: to me `AccThreadSeq` read like `::alpaka::AccCpuSerial`, I thought it was a fixed alias to a single alpaka backend.

jkelling commented on include/pmacc/lockstep/WorkerCfg.hpp:
```diff
@@ -0,0 +1,117 @@
+        {
+            auto const localThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc);
+            auto const blockExtent = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc);
+            auto const linearThreadIdx = alpaka::mapIdx<1u>(localThreadIdx, blockExtent)[0];
+            return Worker<T_Acc, T_numWorkers>(acc, linearThreadIdx);
```
There is no check, including in `Worker`, whether `T_numWorkers` actually matches `blockExtent.prod()`. Routine runtime check here would be feasible, but there should at least be an assertion for debug builds somewhere, maybe an `ALPAKA_ASSERT_OFFLOAD(...)`.

psychocoderHPC replied:
Yes I also thought about it. I will use your suggestion. The check should only be enabled if we compile in debug mode else the performance will be bad and `assert` on GPU is requiring registers what is bad too. For a debug compile this is fine.

psychocoderHPC replied:
I added the check.

psychocoderHPC committed:
refactor lockstep programming

- change interfaces for lockstep programming
- introduce new macro to launche a lockstep kernel
`PMACC_LOCKSTEP_KERNEL(...)`

psychocoderHPC committed:
documentation: update lockstep programming interface usage

psychocoderHPC committed:
example game of life: fix and use new lockstep interfaces

- fix bugs and warnings
- use new lockstep programming interfaces

psychocoderHPC committed:
PMacc tests: use new lockstep programming interfaces

psychocoderHPC committed:
PMacc cuSTL: use new lockstep programming interfaces

psychocoderHPC committed:
PMacc: use new lockstep programming interfaces

Within PMacc the context type `lockstep::Worker` is used instead of the
alpaka accelerator. alpaka accelerator is only required if a cupla/
mallocMC or alpaka method is called.

psychocoderHPC committed:
PIConGPU plugins: use new lockstep programming interfaces

Update plugins to us ethe new interfaces for lockstep programming.

jkelling commented on include/pmacc/lockstep/Worker.hpp:
```diff
@@ -24,53 +24,82 @@
+        /** synchronize all workers
+         *
+         * @attention It is not allowed to call this method in side of a if branch or a loop if it is not guaranteed
+         * that all worker executing the same code branch.
```
```suggestion
         * @attention It is not allowed to call this method inside of an if branch or a loop if it is not guaranteed
         * that all workers are executing the same code branch.
```

jkelling commented on include/pmacc/lockstep/Worker.hpp:
```diff
@@ -24,53 +24,82 @@
+         */
+        HDINLINE static constexpr uint32_t getNumWorkers()
+        {
+            return T_numSuggestedWorkers;
```
Why does this not return `numWorkers`? The caller may know what they suggested already, but is `numWorkers` not the number actually used?

psychocoderHPC replied:
This is a BUG and you are right! This must be `numWorkers`

jkelling commented on share/pmacc/examples/gameOfLife2D/CMakeLists.txt:
```diff
@@ -135,7 +135,7 @@ cupla_add_executable(gameOfLife
 target_link_libraries(gameOfLife PRIVATE pmacc::pmacc)
 if(PNGwriter_FOUND)
-    target_link_libraries(PNGwriter::PNGwriter)
+    target_link_libraries(gameOfLife PRIVATE PNGwriter::PNGwriter)
```
This looks unrelated to this PR... are including this and some other changes to `gameOfLive` because you touched it anyway for the `acc` -> `worker` change? I can see the old code in places cause warnings, too...

psychocoderHPC replied:
Yes you are right. The problem is that this code is not CI validated and I would avoid the large refactoring is blocking all other PR to long.

psychocoderHPC replied:
let me check if it is easy to move the patch into a separate PR.

psychocoderHPC replied:
opened PR #4338

jkelling commented on docs/source/prgpatterns/lockstep.rst:
```diff
@@ -62,11 +62,9 @@ A context variable must be defined outside of ``ForEach`` and should be accessed
 
-    // assume one dimensional indexing of threads within a block
-    uint32_t const workerIdx = cupla::threadIdx(acc).x;
+    // variable 'worker' is provides by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```
```suggestion
    // variable 'worker' is provided by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```

jkelling commented on docs/source/prgpatterns/lockstep.rst:
```diff
@@ -94,11 +90,9 @@ A context variable must be defined outside of ``ForEach`` and should be accessed
 
-    // assume one dimensional indexing of threads within a block
-    uint32_t const workerIdx = cupla::threadIdx(acc).x;
+    // variable 'worker' is provides by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```
```suggestion
    // variable 'worker' is provided by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```

jkelling commented on docs/source/prgpatterns/lockstep.rst:
```diff
@@ -146,14 +138,13 @@ Collective Loop
 
 .. code-block:: bash
 
+    // variable 'worker' is provides by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```
```suggestion
    // variable 'worker' is provided by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```

jkelling commented on docs/source/prgpatterns/lockstep.rst:
```diff
@@ -175,11 +166,9 @@ Non-Collective Loop
 
-    // assume one dimensional indexing of threads within a block
-    uint32_t const workerIdx = cupla::threadIdx(acc).x;
+    // variable 'worker' is provides by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```
```suggestion
    // variable 'worker' is provided by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```

jkelling commented on docs/source/prgpatterns/lockstep.rst:
```diff
@@ -205,9 +194,8 @@ Using a Master Worker
-    // assume one dimensional indexing of threads within a block
-    uint32_t const workerIdx = cupla::threadIdx(acc).x;
-    auto onlyMaster = lockstep::makeMaster(workerIdx);
+    // variable 'worker' is provides by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```
```suggestion
    // variable 'worker' is provided by pmacc if the kernel launch macro `PMACC_LOCKSTEP_KERNEL()` is used.
```

jkelling commented on include/picongpu/fields/currentDeposition/Cache.hpp:
```diff
@@ -46,47 +46,38 @@ namespace picongpu
                 {
                     typename T_Strategy::GridReductionOp const op;
-                    ThreadCollective<T_BlockDescription, T_numWorkers> collectiveAdd(workerIdx);
+                    auto collectiveAdd = makeThreadCollective<T_BlockDescription>();
```
Why is the functor instance called `collectiveAdd`? The operation `op` is supplied by the template parameter `T_Strategy` and the use type name does not suggest anything beyond, that the operation is a reduction. The documentation of this class also does not specify, that the reduction mus be an addition. Is this a misnomer?

psychocoderHPC replied:
I assume it was in the very past always an `Add` and after applying strategies to this part of code not changed.

jkelling replied:
Can we remove the `Add` the? Or maybe replace it be `Reduction`. This is unrelated to this PR. But due to the small size of this change, it would never change otherwise, so this seems to be a good opportunity to fix it.

psychocoderHPC replied:
I will do it with the rebase against https://github.com/ComputationalRadiationPhysics/picongpu/pull/4338

jkelling replied:
It looks like the `Add` is still here...

psychocoderHPC replied:
changed to `collectiveFlush`

jkelling commented on include/picongpu/particles/flylite/NonLTE.tpp:
```diff
@@ -29,7 +29,7 @@
 #include <pmacc/Environment.hpp>
 #include <pmacc/dataManagement/ISimulationData.hpp>
-#include <pmacc/traits/GetNumWorkers.hpp>
+#include <pmacc/lockstep/lockstep.hpp>
```
Is this include even required here? Considering there are no other changes in this file, it could that `GetNumWorkers.hpp` was also superfluous.

psychocoderHPC replied:
I removed it, it was wrong before and auto substitution changed it.

psychocoderHPC committed:
WokrerCfg: kernel launch parameter validation

Add check for debugging to enshure that the number of threads in a block
fits the number of workers used for lockstep programming.

psychocoderHPC committed:
PIConGPU collision: use new lockstep programming interfaces

psychocoderHPC committed:
PIConGPU core: use new lockstep programming interfaces

psychocoderHPC committed:
apply Jeffry's github review suggestions

pordyna reviewed (dismissed):
I looked over the  `PIConGPU collision: use new lockstep programming interfaces` commit, and it looks fine.

PrometheusPi commented:
@psychocoderHPC https://github.com/ComputationalRadiationPhysics/picongpu/pull/4338 was merged yesterday, you can rebase.

psychocoderHPC commented:
PR is rebased and should pass the CI now

jkelling commented on include/picongpu/particles/collision/detail/CollisionContext.hpp:
```diff
@@ -29,21 +29,23 @@ namespace picongpu
                 struct CollisionContext
                 {
-                    T_Acc const* m_acc;
+                    T_Worker const* m_worker;
```
I would consider making this member constant, for the benefit of the compiler, so this looks more like a stored reference. ... unless you intend to assign to this type.

jkelling replied:
.. on second though: It just be a reference then. Both variants disable assignment and move ... I doubt than either is relevant with this context-type.

Anyways: this is unrelated to the PR, hence optional.

psychocoderHPC replied:
This I would keep for now as it is because this is tested. Changing this could break thinks and postpone this PR to much

jkelling commented on include/picongpu/particles/functor/misc/Rng.hpp:
```diff
@@ -51,7 +51,7 @@ namespace picongpu
                     using Distribution = T_Distribution;
                     using RNGFactory = pmacc::random::RNGProvider<simDim, random::Generator>;
                     using RngHandle = typename RNGFactory::Handle;
-                    using RandomGen = RngWrapper<cupla::Acc, typename RngHandle::GetRandomType<Distribution>::type>;
+
 
                     /** constructor
```
Not, sure, but I think there should only be one blank line here.

psychocoderHPC replied:
removed the double blank line
